File: bench/sieve.c
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    long limit = 2000000;
    long *flags = malloc((size_t)(limit + 1) * sizeof(long));
    for (long i = 0; i <= limit; i++) flags[i] = 1;

    for (long p = 2; p * p <= limit; p++)
        if (flags[p] == 1)
            for (long m = p * p; m <= limit; m += p) flags[m] = 0;

    long count = 0;
    for (long n = 2; n <= limit; n++) count += flags[n];

    printf("%ld\n", count);
    free(flags);
    return 0;
}

File: bench/sieve.zig
const std = @import("std");

fn build(gpa: std.mem.Allocator, limit: i64) !std.ArrayList(i64) {
    var flags: std.ArrayList(i64) = .empty;
    var i: i64 = 0;
    while (i <= limit) : (i += 1) try flags.append(gpa, 1);

    var p: i64 = 2;
    while (p * p <= limit) : (p += 1) {
        if (flags.items[@intCast(p)] == 1) {
            var m = p * p;
            while (m <= limit) : (m += p) flags.items[@intCast(m)] = 0;
        }
    }
    return flags;
}

pub fn main(init: std.process.Init) !void {
    const limit: i64 = 2000000;
    var flags = try build(init.gpa, limit);
    defer flags.deinit(init.gpa);

    var count: i64 = 0;
    var n: i64 = 2;
    while (n <= limit) : (n += 1) {
        if (flags.items[@intCast(n)] == 1) count += 1;
    }
    std.debug.print("{d}\n", .{count});
}
